File: core/src/commonMain/kotlin/misskey4j/internal/api/ListsResourceImpl.java
package misskey4j.internal.api;

import javax.annotation.Nonnull;

import misskey4j.MisskeyAPI;
import misskey4j.api.ListsResource;
import misskey4j.api.request.lists.UsersListsCreateRequest;
import misskey4j.api.request.lists.UsersListsDeleteRequest;
import misskey4j.api.request.lists.UsersListsListRequest;
import misskey4j.api.request.lists.UsersListsPullRequest;
import misskey4j.api.request.lists.UsersListsPushRequest;
import misskey4j.api.request.lists.UsersListsShowRequest;
import misskey4j.api.request.lists.UsersListsUpdateRequest;
import misskey4j.api.response.lists.UsersListsListResponse;
import misskey4j.api.response.lists.UsersListsShowResponse;
import misskey4j.entity.NoteList;
import misskey4j.entity.share.Response;

public class ListsResourceImpl extends AbstractResourceImpl implements ListsResource {

    public ListsResourceImpl(String uri, String i) {
        super(uri, i);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public Response<UsersListsListResponse[]> list(
            UsersListsListRequest request) {
        return post(UsersListsListResponse[].class,
                MisskeyAPI.Lists.code(), request);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public Response<UsersListsShowResponse> show(
            UsersListsShowRequest request) {
        return post(UsersListsShowResponse.class,
                MisskeyAPI.ListsShow.code(), request);
    }

    @Override
    public void push(@Nonnull String listId, @Nonnull String userId) {

        UsersListsPushRequest request = UsersListsPushRequest.builder()
                        .userId(userId)
                        .listId(listId)
                        .build();
        post(MisskeyAPI.ListsPush.code(), request);
    }

    @Override
    public void pull(@Nonnull String listId, @Nonnull String userId) {

        UsersListsPullRequest request = UsersListsPullRequest.builder()
                        .userId(userId)
                        .listId(listId)
                        .build();
        post(MisskeyAPI.ListsPull.code(), request);
    }

    @Override
    public Response<NoteList> create(@Nonnull String name) {

        UsersListsCreateRequest request = new UsersListsCreateRequest(name);
        return post(NoteList.class, MisskeyAPI.ListsCreate.code(), request);
    }

    @Override
    public Response<NoteList> update(@Nonnull String listId, @Nonnull String name) {

        UsersListsUpdateRequest request = new UsersListsUpdateRequest(listId, name);
        return post(NoteList.class, MisskeyAPI.ListsUpdate.code(), request);
    }

    @Override
    public void delete(@Nonnull String listId) {

        UsersListsDeleteRequest request = new UsersListsDeleteRequest(listId);
        post(MisskeyAPI.ListsDelete.code(), request);
    }

}

File: core/src/commonMain/kotlin/misskey4j/internal/api/ListsResourceImpl.kt
package misskey4j.internal.api

import misskey4j.MisskeyAPI.*
import misskey4j.api.ListsResource
import misskey4j.api.request.lists.*
import misskey4j.api.response.lists.UsersListsListResponse
import misskey4j.api.response.lists.UsersListsShowResponse
import misskey4j.entity.NoteList
import misskey4j.entity.share.Response

class ListsResourceImpl(
    uri: String,
    i: String
) : AbstractResourceImpl(uri, i),
    ListsResource {

    /**
     * {@inheritDoc}
     */
    override fun list(
        request: UsersListsListRequest
    ): Response<Array<UsersListsListResponse>> {
        return post(Lists.path, request)
    }

    /**
     * {@inheritDoc}
     */
    override fun show(
        request: UsersListsShowRequest
    ): Response<UsersListsShowResponse> {
        return post(ListsShow.path, request)
    }

    /**
     * {@inheritDoc}
     */
    override fun push(
        request: UsersListsPushRequest
    ): Response<Unit> {
        return post(ListsPush.path, request)
    }

    /**
     * {@inheritDoc}
     */
    override fun pull(
        request: UsersListsPullRequest
    ): Response<Unit> {
        return post(ListsPull.path, request)
    }

    /**
     * {@inheritDoc}
     */
    override fun create(
        request: UsersListsCreateRequest
    ): Response<NoteList> {
        return post(ListsCreate.path, request)
    }

    /**
     * {@inheritDoc}
     */
    override fun update(
        request: UsersListsUpdateRequest
    ): Response<NoteList> {
        return post(ListsUpdate.path, request)
    }

    /**
     * {@inheritDoc}
     */
    override fun delete(
        request: UsersListsDeleteRequest
    ): Response<Unit> {
        return post(ListsDelete.path, request)
    }
}
